Add lint writing documentation

phansch opened the pull request:
[Rendered](https://github.com/phansch/rust-clippy/blob/adding_lints/doc/adding_lints.md)

This adds a new documentation page that explains how to write Clippy
lints. It guides the reader through creating a `foo` function lint.

I plan to iterate a bit more on the prose of some sections, but I think the
general structure is fine now, so I'm looking forward to feedback =)

One thing I'm not sure about: I felt like this is too big for CONTRIBUTING.md
so I put it into a new `doc/` directory. I can imagine having more
documentation in the future, so we might even want to create a book using
mdbook instead? Or should everything go into CONTRIBUTING.md?

Further things left to do:

- [x] Link from CONTRIBUTING.md
- [x] Remove things covered in this guide from CONTRIBUTING.md
- [x] Section about `clippy::author` attribute
- [x] Run `remark-lint` on CI over the `doc` directory and fix things

Manishearth commented:
This looks really good so far, thanks for doing this!!

As an aside: We should probably build a longer version of the lint naming conventions, we have a bunch of our own (like `useless_foo`). https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints is pretty basic

detrumi commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,324 @@
+have to deal with the AST and don't have to deal with the type system at all.
+This is good, because it makes writing this particular lint less complicated.
+
+We have to make this decision with every new Clippy lint. It boils down to using either `EarlyLintPass` or `LateLintPass`. This is a result of Rust's compilation process. You can read more about it [in the rustc guide][compilation_stages].
```
You link to the rustc guide, but that page doesn't mention early/late. Maybe mention in which stage the early/late lints are run?

phansch replied:
You're right, I think I will leave that last sentence out for now. There's not really a place where EarlyLintPass and LateLintPass are explained apart from this PR.

detrumi commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,324 @@
+        FnKind::ItemFn(ident, ..) | FnKind::Method(ident, ..) => {
+            return ident.name == "foo"
+        },
+        FnKind::Closure(..) => return false
```
```suggestion
        FnKind::Closure(..) => false
```

detrumi commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,324 @@
+fn is_foo_fn(fn_kind: FnKind<'_>) -> bool {
+    match fn_kind {
+        FnKind::ItemFn(ident, ..) | FnKind::Method(ident, ..) => {
+            return ident.name == "foo"
```
```suggestion
            ident.name == "foo"
```

flip1995 commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,366 @@
+impl EarlyLintPass for Pass {
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, fn_kind: FnKind<'_>, _: &FnDecl, span: Span, _: NodeId) {
+        if !is_foo_fn(fn_kind) { return; }
+        span_help_and_lint(
```
Bikeshed: Personally I would prefer to write
```rust
if is_foo_fn(fn_kind) {
    span_help_and_lint(..);
}
```
instead of an early return. This is easier to read IMHO. "If is_foo_function then lint" vs "If it's not a foo function then return otherwise lint".

flip1995 commented on doc/adding_lints.md:
````diff
@@ -0,0 +1,366 @@
+}
+```
+
+Now you'll want to also run the full test suite with `cargo test`. Apart from
````
I would add another sentence here:

"At this point running `cargo uitest` on your test file should produce the expected output (don't forget to run the `update-all-references.sh` script [Maybe mention the script here again, I don't know...]). Now you'll want..."

flip1995 commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,366 @@
+* [`in_macro`][in_macro] and [`in_external_macro`][in_external_macro]
+* [`Span`][span]
+* [Clippy diagnostics][diagnostics]
+* [`Applicability`][applicability]
```
Maybe also mention the `utils` module here. There are many helpful convenience functions ready to use. (`snippet`, `Sugg`, `is_expn_of`, ...)

phansch replied:
Done in https://github.com/rust-lang/rust-clippy/pull/3824/commits/31b62841e0252e3d33e10e9b3f39b9b7e2c8f63d

phansch commented:
Thanks for the reviews so far! I added a small TODO list to the PR description. I will fix those and give it another read/try later today. 

> In the future I would like a section "Advanced Lints", where some other useful things are mentioned and explained.  I'm thinking of:  Visitor, LateLintPass example, PreExpansionPass, Constants, MIR-Lints

These are all good ideas! I hope having the documentation in it's own directory lowers the bar a tiny bit for new documentation. I found that working with the existing CONTRIBUTING.md was getting a bit unpleasant.

phansch commented:
I didn't have time to re-read the whole thing yet and try it out, but I think I addressed all the remaining issues, including the cleanup of `CONTRIBUTING.md`.

oli-obk commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,450 @@
+
+Clippy uses UI tests for testing. UI tests check that the output of Clippy is
+exactly as expected. Each test is just a plain Rust file that contains the code
+we want to check. The output of Clippy is compared against a `.stderr` file.
```
"Note that you don't have to create this yourself, we'll get to generating `.stderr` files further down"

phansch replied:
Added in https://github.com/rust-lang/rust-clippy/pull/3824/commits/e9d37ca79a7e457a664a032e736be2eb9b9af3c7

oli-obk commented on doc/adding_lints.md:
```diff
@@ -0,0 +1,450 @@
+lint to generate Clippy code that detects the offending pattern. It does not
+work for all of the Rust syntax, but can give a good starting point.
+
+First, create a new UI test file in the `tests/ui/` directory with the pattern
```
I suggest to just do this on the playground, as that requires zero setup.

flip1995 commented on doc/adding_lints.md:
````diff
@@ -309,6 +373,19 @@ declare_clippy_lint! { /* ... */ }
+rustup component add rustfmt
+```
+
+Use `cargo fmt --all` to format the whole codebase.
````
I recently realized, that this doesn't format the test files. 

The workaround in travis is to walk the `tests` folder and call `rustfmt` on each test file:
https://github.com/rust-lang/rust-clippy/blob/9f2055637e3a32064517c888f7d69b4fad6d67c5/ci/base-tests.sh#L62-L68

Maybe we should provide a script which does just that.

```bash
cargo fmt --all
for file in `find tests -not -path "tests/ui/doc.rs" -not -path "tests/ui/unused_unit.rs" | grep "\.rs$"` ; do 
    rustfmt ${file}
done
```

phansch replied:
Yeah, that's something that would be super nice! I created https://github.com/rust-lang/rust-clippy/issues/3835 to further track this.

felix91gr commented:
Whaaat! This PR is awesome 😍 there's so much info I learned this last week as I got started, that you've neatly put together here. Thanks! This will surely help a lot of people 😊

felix91gr commented:
Have you seen [this](https://github.com/rust-lang/rust-clippy/projects/2)? Maybe it'd be nice to connect it with this PR :D

phansch commented:
> Have you seen [this](https://github.com/rust-lang/rust-clippy/projects/2)? Maybe it'd be nice to connect it with this PR :D

Yup, I just cleaned that up a bit and added a 'Done' column and created https://github.com/rust-lang/rust-clippy/issues/3843.

Apart from that, I think this PR is ready for a final review. The rest can be done in subsequent PRs.

phansch commented on doc/adding_lints.md:
````diff
@@ -0,0 +1,445 @@
+Please document your lint with a doc comment akin to the following:
+
+```rust
+/// **What it does:** Checks for ... (describe what the lint matches).
````
Need to update this due to #3845

phansch commented:
Going to go ahead and get this merged then, we can still improve it going forward ✨

@bors r+

bors commented:
:pushpin: Commit 34685a5f6039ec406962730cf2613044a4d8f31c has been approved by `phansch`

<!-- @bors r=phansch 34685a5f6039ec406962730cf2613044a4d8f31c -->

bors commented:
:hourglass: Testing commit 34685a5f6039ec406962730cf2613044a4d8f31c with merge 920112d723e3d4162e72ad29070ebf9cf65feb38...
<!-- homu: {"type":"BuildStarted","head_sha":"34685a5f6039ec406962730cf2613044a4d8f31c","merge_sha":"920112d723e3d4162e72ad29070ebf9cf65feb38"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/103779157), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/22950013)
Approved by: phansch
Pushing 920112d723e3d4162e72ad29070ebf9cf65feb38 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"phansch","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/22950013","checks-travis":"https://travis-ci.com/rust-lang/rust-clippy/builds/103779157"},"merge_sha":"920112d723e3d4162e72ad29070ebf9cf65feb38"} -->
